task_sdk/src/airflow/sdk/execution_time/supervisor.py

```diff
@@ -0,0 +1,552 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Supervise and run Tasks in a subprocess."""
+
+from __future__ import annotations
+
+import atexit
+import io
+import logging
+import os
+import selectors
+import signal
+import sys
+import time
+import weakref
+from collections.abc import Generator
+from contextlib import suppress
+from datetime import datetime, timezone
+from socket import socket, socketpair
+from typing import TYPE_CHECKING, BinaryIO, Callable, ClassVar, Literal, NoReturn, cast, overload
+from uuid import UUID
+
+import attrs
+import httpx
+import msgspec
+import psutil
+import structlog
+
+from airflow.sdk.api.client import Client
+from airflow.sdk.api.datamodels._generated import TaskInstanceState
+from airflow.sdk.execution_time.comms import ConnectionResponse, GetConnection, StartupDetails, ToSupervisor
+
+if TYPE_CHECKING:
+    from structlog.typing import FilteringBoundLogger
+
+    from airflow.sdk.api.datamodels.activities import ExecuteTaskActivity
+    from airflow.sdk.api.datamodels.ti import TaskInstance
+
+
+__all__ = ["WatchedSubprocess", "supervise"]
+
+log: FilteringBoundLogger = structlog.get_logger(logger_name="supervisor")
+
+# TODO: Pull this from config
+SLOWEST_HEARTBEAT_INTERVAL: int = 30
+# Don't heartbeat more often than this
+FASTEST_HEARTBEAT_INTERVAL: int = 5
+
+
+@overload
+def mkpipe() -> tuple[socket, socket]: ...
+
+
+@overload
+def mkpipe(remote_read: Literal[True]) -> tuple[socket, BinaryIO]: ...
+
+
+def mkpipe(
+    remote_read: bool = False,
+) -> tuple[socket, socket | BinaryIO]:
+    """
+    Create a pair of connected sockets.
+
+    The inheritable flag will be set correctly so that the end destined for the subprocess is kept open but
+    the end for this process is closed automatically by the OS.
+    """
+    rsock, wsock = socketpair()
+    local, remote = (wsock, rsock) if remote_read else (rsock, wsock)
+
+    remote.set_inheritable(True)
+    local.setblocking(False)
+
+    io: BinaryIO | socket
+    if remote_read:
+        # If _we_ are writing, we don't want to buffer
+        io = cast(BinaryIO, local.makefile("wb", buffering=0))
+    else:
+        io = local
+
+    return remote, io
+
+
+def _subprocess_main():
+    from airflow.sdk.execution_time.task_runner import main
+
+    main()
+
+
+def _fork_main(
+    child_stdin: socket,
+    child_stdout: socket,
+    child_stderr: socket,
+    log_fd: int,
+    target: Callable[[], None],
+) -> NoReturn:
```

Comment:
Worth adding docstring with the tldr of what things are done in this func

Reply:
Doc string, and refactor to small name functions to make flow a bit clearer hopefully https://github.com/apache/airflow/pull/43893/commits/c54c51e8257f87410267aa5da636a326a9028089#diff-c2651fdee1a25e091e2a9d4f937f8032ca3d289d0de76f38ed88aee5df0f880dR159-R183